internal/stanza/emitter.go

```diff
@@ -17,46 +17,170 @@ package stanza // import "github.com/open-telemetry/opentelemetry-collector-cont
+	return LogEmitterOption(func(le *LogEmitter) {
+		le.flushInterval = flushInterval
+	})
+}
+
+// LogEmitterWithLogger returns an option that makes the LogEmitter use the specified logger
+func LogEmitterWithLogger(logger *zap.SugaredLogger) LogEmitterOption {
+	return LogEmitterOption(func(le *LogEmitter) {
+		le.OutputOperator.BasicOperator.SugaredLogger = logger
+	})
 }
 
 // NewLogEmitter creates a new receiver output
-func NewLogEmitter(logger *zap.SugaredLogger) *LogEmitter {
-	return &LogEmitter{
+func NewLogEmitter(opts ...LogEmitterOption) *LogEmitter {
+	le := &LogEmitter{
 		OutputOperator: helper.OutputOperator{
 			BasicOperator: helper.BasicOperator{
 				OperatorID:    "log_emitter",
 				OperatorType:  "log_emitter",
-				SugaredLogger: logger,
+				SugaredLogger: zap.NewNop().Sugar(),
 			},
 		},
-		logChan: make(chan *entry.Entry),
+		logChan:       make(chan []*entry.Entry),
+		maxBatchSize:  defaultMaxBatchSize,
+		batch:         make([]*entry.Entry, 0, defaultMaxBatchSize),
+		flushInterval: defaultFlushInterval,
+		cancel:        func() {},
 	}
-}
 
-// Process will emit an entry to the output channel
-func (e *LogEmitter) Process(ctx context.Context, ent *entry.Entry) error {
-	select {
-	case e.logChan <- ent:
-	case <-ctx.Done():
+	for _, opt := range opts {
+		opt(le)
 	}
+
+	return le
+}
+
+// Start starts the goroutine(s) required for this operator
+func (e *LogEmitter) Start(_ operator.Persister) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	e.cancel = cancel
+
+	e.wg.Add(1)
+	go e.flusher(ctx)
 	return nil
 }
 
-// Stop will close the log channel
+// Stop will close the log channel and stop running goroutines
 func (e *LogEmitter) Stop() error {
 	e.stopOnce.Do(func() {
+		e.cancel()
+		e.wg.Wait()
+
 		close(e.logChan)
 	})
+
+	return nil
+}
+
+// Process will emit an entry to the output channel
+func (e *LogEmitter) Process(ctx context.Context, ent *entry.Entry) error {
+	if batch := e.appendEntry(ent); batch != nil {
+		e.flush(ctx, batch)
+	}
+
+	return nil
+}
+
+// appendEntry appends the entry to the current batch. If maxBatchSize is reached, a new batch will be made, and the old batch
+// (which should be flushed) will be returned
+func (e *LogEmitter) appendEntry(ent *entry.Entry) []*entry.Entry {
+	e.batchMux.Lock()
+	defer e.batchMux.Unlock()
+
+	e.batch = append(e.batch, ent)
+	if uint(len(e.batch)) >= e.maxBatchSize {
+		// maxBatchSize has been exceeded; create a new batch, returning the old one.
+		return e.makeNewBatchNoLock()
+	}
+
 	return nil
 }
+
+// flusher flushes the current batch every flush interval. Intended to be run as a goroutine
+func (e *LogEmitter) flusher(ctx context.Context) {
+	defer e.wg.Done()
+
+	ticker := time.NewTicker(e.flushInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			if batch := e.makeNewBatch(); batch != nil {
+				e.flush(ctx, batch)
+			}
```

Comment:
I keep finding these "makeNewBatch" functions difficult to reason about. 

`makeNewBatchNoLock` is really only deduplicating a couple simple lines, so seems unnecessary. I think we ought to go a step further and roll all of this back into the callers. Here for example:
  ```suggestion
				e.batchMux.Lock()			
				if len(e.batch) > 0 {
					oldBatch, e.batch := e.batch, make([]*entry.Entry, 0, e.maxBatchSize)
					e.batchMux.Unlock()
					e.flush(ctx, oldBatch)
				}
```

Reply:
The reason why I have these functions is to make it easy to reason about where the critical section begins/ends (critical sections always begin and end at function level). 

makeNewBatchNoLock could certainly be folded in now that you mention it, it's basically one line. I mainly just want to avoid have unlocks in branches, so that one `Lock` call has multiple possible `Unlocks`; I feel dicey doing that.

I'll try refactoring with that in mind.

Reply:
I ironed out makeNewBatch, makeNewBatchNoLock, and appendEntry. Hopefully it reads better now.

Reply:
I tend to like having `Lock/Unlock` in a function too, and wouldn't oppose it if you wanted to keep those. Getting rid of `makeNewBatchNoLock` was the main thing.

Reply:
Cool, I like the function-level locking, so I added makeNewBatch and appendEntry back in.